hare3/hare_test.go

```diff
@@ -505,37 +508,38 @@ type testTracer struct {
 	sent        chan *Message
 }
 
-func (t *testTracer) waitStopped() types.LayerID {
-	wait := 10 * time.Second
+func waitForChan[T any](t testing.TB, ch <-chan T, timeout time.Duration, failureMsg string) T {
+	var value T
 	select {
-	case <-time.After(wait):
-		require.FailNow(t, "didn't stop", "wait %v", wait)
-	case lid := <-t.stopped:
-		return lid
+	case <-time.After(timeout):
+		builder := strings.Builder{}
+		pprof.Lookup("goroutine").WriteTo(&builder, 2)
+		t.Fatalf(failureMsg+", waited: %v, stacktraces:\n%s", timeout, builder.String())
```

Comment:
```suggestion
		var buf bytes.Buffer
		pprof.Lookup("goroutine").WriteTo(&buf, 2)
		t.Fatalf(failureMsg+", waited: %v, stacktraces:\n%s", timeout, buf.String())
```

Reply:
What's the difference?

Reply:
`bytes.Buffer` is faster when `Write` is used to append to the buffer, `strings.Builder` is faster when appending strings; `WriteTo` uses `Write` not

```go
func BenchmarkBuffer(b *testing.B) {
	data := make([]byte, 1024)
	rand.Read(data)

	var buf bytes.Buffer
	var sb strings.Builder

	b.Run("bytes.Buffer", func(b *testing.B) {
		for range b.N {
			buf.Write(data)
		}
	})

	b.Run("strings.Builder", func(b *testing.B) {
		for range b.N {
			sb.Write(data)
		}
	})
}
```

```
goos: linux
goarch: arm64
pkg: github.com/spacemeshos/go-spacemesh/fetch
BenchmarkBuffer/bytes.Buffer-10         	 1000000	      1909 ns/op	    2113 B/op	       0 allocs/op
BenchmarkBuffer/strings.Builder-10      	  829036	      7350 ns/op	    5472 B/op	       0 allocs/op
PASS
coverage: 0.1% of statements
ok  	github.com/spacemeshos/go-spacemesh/fetch	8.146s
```

Reply:
First, let's make this benchmark measure the same thing as the code.

```go
func BenchmarkBuffer(b *testing.B) {
	lookup := pprof.Lookup("goroutine")

	b.Run("bytes.Buffer", func(b *testing.B) {
		var buffer bytes.Buffer
		for range b.N {
			lookup.WriteTo(&buffer, 2)
		}
	})

	b.Run("strings.Builder", func(b *testing.B) {
		var sb strings.Builder
		for range b.N {
			lookup.WriteTo(&sb, 2)
		}
	})
}
```

The speed of both approaches is similar:
```
goos: linux
goarch: amd64
pkg: github.com/spacemeshos/go-spacemesh/hare3
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkBuffer/bytes.Buffer-20         	   10846	    106467 ns/op	 1064044 B/op	       1 allocs/op
BenchmarkBuffer/strings.Builder-20      	   10069	    115654 ns/op	 1070063 B/op	       1 allocs/op
PASS
```

Reply:
Performance is absolutely not important here :shrug:. 

> I think you can also just write directly to os.Stdout

I didn't want to write directly to stdout to avoid all sorts of weird behaviors if multiple tests run in parallel.